core/src/main/java/nl/nn/adapterframework/soap/SoapWrapper.java

```diff
@@ -267,34 +270,48 @@ public Message signMessage(Message soapMessage, String user, String password, bo
 			Document doc = unsignedEnvelope.getOwnerDocument();
 
 			// create security header and insert it into unsigned envelope
-			WSSecHeader secHeader = new WSSecHeader();
-			secHeader.insertSecurityHeader(doc);
+			WSSecHeader secHeader = new WSSecHeader(doc);
+			secHeader.insertSecurityHeader();
 
 			// add a UsernameToken
-			WSSecUsernameToken tokenBuilder = new WSSecUsernameToken();
+			WSSecUsernameToken tokenBuilder = new WSSecUsernameToken(secHeader);
+			tokenBuilder.setIdAllocator(idAllocator);
 			if (passwordDigest) {
 				tokenBuilder.setPasswordType(WSConstants.PASSWORD_DIGEST);
 			} else {
 				tokenBuilder.setPasswordType(WSConstants.PASSWORD_TEXT);
 			}
+			tokenBuilder.setPrecisionInMilliSeconds(false);
 			tokenBuilder.setUserInfo(user, password);
+			WSTimeSource timesource = tokenBuilder.getWsTimeSource();
 			tokenBuilder.addNonce();
 			tokenBuilder.addCreated();
-			tokenBuilder.prepare(doc);
-
-			WSSecSignature sign = new WSSecSignature();
-			sign.setUsernameToken(tokenBuilder);
-			sign.setKeyIdentifierType(WSConstants.UT_SIGNING);
-			sign.setSignatureAlgorithm(XMLSignature.ALGO_ID_MAC_HMAC_SHA1);
-			sign.build(doc, null, secHeader);
-
-			tokenBuilder.prependToHeader(secHeader);
+			tokenBuilder.prepare(null);
+			Element element = tokenBuilder.getUsernameTokenElement();
+			String nonce = XmlUtils.getChildTagAsString(element, "wsse:Nonce");
+			byte[] decodedNonce = org.apache.xml.security.utils.XMLUtils.decode(nonce);
+			String created = XmlUtils.getChildTagAsString(element, "wsu:Created");
```

Comment:
Moet het niet de 'andere partij' zijn die de nonce aanlevert, bij wijze van challenge?

Reply:
Nee, het zit raar in elkaar, maar de nonce wordt berekend, en middels de nonce, een timestamp en password wordt een signature berekend, en die wordt gebruikt voor de challenge.